src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,459 @@ namespace System
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            int length = ClampEnd(span, start, trimElement);
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int length = ClampEnd(span, 0, trimElement);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampStart<T>(in ReadOnlySpan<T> span, in T trimElement)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty)
+                return 0;
+
+            int start = 0;
+
+            if (trimElement == null)
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (span[start] != null)
+                        break;
+                }
+            }
+            else
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (span[start] != null && !span[start].Equals(trimElement))
+                        break;
+                }
+            }
+
+            return start;
+        }
+
+        /// <summary>
+        /// Delimits all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampEnd<T>(in ReadOnlySpan<T> span, int start, in T trimElement)
+            where T : IEquatable<T>
+        {
+            Debug.Assert((uint)start < span.Length);
```

Comment:
Couldn't start also == Length?

Reply:
Yes, when `start` and `length` are both `0` - I had previously mitigated that call path hitting the helper via the `IsEmpty` check. So I need the `0` case now.
Is there another circumstance that equality could be true?
(I am debating whether I use `start <= len` or `start == 0 || start < len`)

Reply:
> Is there another circumstance that equality could be true?

Presumably if the whole span was trimmed away as part of the clamp start?

Reply:
Ah, of course. Fixed

Reply:
Considering your point, do you think it's worth adding this check into the callsite(s) instead?
```csharp
        public static Memory<T> Trim<T>(this Memory<T> memory, T trimElement)
            where T : IEquatable<T>
        {
// Existing
            Span<T> span = memory.Span;
            int start = ClampStart(span, trimElement);
// Add this guard?
            if (start >= memory.Length)
                return Memory<T>.Empty;
// Existing
            int length = ClampEnd(span, start, trimElement);
            return memory.Slice(start, length);
        }
```

Reply:
Thought about it a bit. The case where the _entire input_ is consumed by `TrimStart` does not seem to be as likely as the case where just a few elements are consumed. So this would be a wasteful additional branch.